src/sparse/kronecker.rs

```diff
@@ -1,6 +1,25 @@
 use crate::indexing::SpIndex;
 use crate::sparse::prelude::*;
 
+/// Trait for types that are valid to compute a Kronecker product.
+/// This includes all classic scalar types, but can be extended to matrices
+/// as well, which will enable computing the kronecker product of sparse
+/// by block matrices.
+pub trait Kronecker<B = Self> {
+    type Output;
+    fn kron(&self, b: &B) -> <Self as Kronecker<B>>::Output;
+}
+
+impl<N> Kronecker<N> for N
+where
+    for<'r> &'r N: std::ops::Mul<&'r N, Output = N>,
+{
+    type Output = N;
+    fn kron(&self, b: &N) -> <Self as Kronecker<N>>::Output {
+        self * b
+    }
+}
+
```

Comment:
I'm afraid this blanket implementation will prevent a block-sparse kronecker product. We should instead use something like the following macro:
```rust
macro_rules! scalar_kronecker {
    ($scalar: ident) => {
        impl Kronecker for $scalar {
            type Output = Self;
            fn kron(&self, b: &Self) -> Self::Output {
                self * b
            }
        }
    };
}

scalar_kronecker!(u8);
scalar_kronecker!(i8);
scalar_kronecker!(u16);
scalar_kronecker!(i16);
scalar_kronecker!(u32);
scalar_kronecker!(i32);
scalar_kronecker!(u64);
scalar_kronecker!(i64);
scalar_kronecker!(isize);
scalar_kronecker!(usize);
scalar_kronecker!(f32);
scalar_kronecker!(f64);
```

Reply:
And the signature of `kronecker_product` should be
```rust
pub fn kronecker_product<
    Nin: Clone + Default,
    Nout: Clone + Default,
    I: SpIndex,
    Iptr: SpIndex,
>(
    mut a: CsMatViewI<Nin, I, Iptr>,
    mut b: CsMatViewI<Nin, I, Iptr>,
) -> CsMatI<Nout, I, Iptr>
where
    Nin: Kronecker<Output = Nout>,
{
```

Reply:
Quite right, I'll change that.

Reply:
Should be better now.

Reply:
Looks perfect!